c/meterpreter/source/extensions/sniffer/sniffer.c

```diff
@@ -38,61 +41,123 @@ Command customCommands[] =
 
-HANDLE hMgr;
-DWORD hErr;
+	struct sockaddr_in srcaddr;
 
-/*!
- * @brief Initialises the packet sniffer SDK.
- * @returns Indication of success or failure.
- * @retval ERROR_ACCESS_DENIED The packet sniffer SDK failed to create a PSSDK manager instance.
- *                             This could be due to insufficient privs (run as root/SYSTEM).
- * @retval ERROR_SUCCESS Initialisation was successful.
- */
-DWORD pktsdk_initialize(void)
+	memset(&srcaddr, 0, sizeof(struct sockaddr_in));
+	memcpy(&(srcaddr.sin_addr), src, sizeof(srcaddr.sin_addr));
+
+	srcaddr.sin_family = af;
+	if (WSAAddressToString((struct sockaddr*) &srcaddr, sizeof(struct sockaddr_in), 0, dst, (LPDWORD)&cnt) != 0) {
+		DWORD rv = WSAGetLastError();
+		return NULL;
+	}
+	return dst;
+}
+
+
+struct iface_entry {
+	unsigned char name;
+	unsigned char hwaddr[6];
+	uint32_t mtu;
+	uint32_t index;
+	unsigned char flags[101];
+	int addr_count;
+	struct iface_address *addr_list;
+};
+
+struct ifaces_list {
+	int entries;
+	struct iface_entry ifaces[0];
+};
+
+char *get_interface_name_by_index(unsigned int fidx)
 {
-	dprintf("sniffer>> calling MgrCreate()...");
+	unsigned int idx;
+	char errbuf[PCAP_ERRBUF_SIZE+4];
+	static char device_name[64];				// PKS, probably safe, due to snifferm mutex
+	struct ifaces_list *ifaces;
+	pcap_if_t *interfaces, *int_iter;
+
+
+	interfaces = int_iter = NULL;
+	ifaces = NULL;
+	idx = 1;
+
+	memset(device_name, 0, sizeof(device_name));
 
-	hMgr = MgrCreate();
-	if(! hMgr)
+	if(pcap_findalldevs(&interfaces, errbuf) == -1)
 	{
-		dprintf("sniffer>> failed to allocate a new Mgr object");
-		hErr = ERROR_ACCESS_DENIED;
-		return(hErr);
+		dprintf("pcap_findalldevs failed, errbuf was : %s", errbuf);
+		return NULL;
+	}
+	else
+	{ //pcap_findalldevs suceeded
+		for(int_iter = interfaces; int_iter; int_iter = int_iter->next)
+		{
+			if(fidx == idx++)
+			{
+				#
+				strncpy(device_name, int_iter->name, sizeof(device_name)-1);
+				break;
+			}
+		}
 	}
 
-	hErr = MgrInitialize(hMgr);
-	if(hErr != HNERR_OK)
+	if (ifaces)
 	{
-		MgrDestroy(hMgr);
-		hMgr = NULL;
+		free(ifaces);
 	}
 
-	dprintf("sniffer>> Mgr object initialized with return %d (handle %d)", hErr, hMgr);
-	return hErr;
+	if (interfaces)
+	{
+		pcap_freealldevs(interfaces);
+	}
+
+	return device_name[0] ? device_name : NULL;
+
 }
 
-HANDLE pktsdk_interface_by_index(unsigned int fidx) {
-	unsigned idx = 1;
-	HANDLE hCfg;
+// http://www.google.com/#q=peter+packet
 
-	dprintf("sniffer>> pktsdk_interface_by_index(%d)", fidx);
+typedef struct PeterPacket
+{
+	struct pcap_pkthdr h;
+	u_char bytes[0];
+} PeterPacket;
+
+char *packet_filter;
+
+#define PktDestroy(x) free((void *)(x))
+#define PktGetPacketSize(x) (((PeterPacket *)(x))->h.caplen)
 
-	hCfg = MgrGetFirstAdapterCfg(hMgr);
-	do {
-		if(fidx == idx++) return hCfg;
-	}while((hCfg = MgrGetNextAdapterCfg(hMgr,hCfg)) != NULL);
-	return NULL;
+DWORD PktGetId(void *handle, DWORD *thi)

```

Comment:
Rambling comment here (TL;DR: I think this is OK as-is, but we might have an existing bug which could result in misordered packets in the final pcap the MSF user downloads)

Due to how pcap calculates timestamp on different systems (note: that statement is made based on empirical observations I've made on different pcap-using projects, I haven't ever audited the pcap code directly or done focused testing), you may not get good (or any at all) "low" resolution, even on 64-bit OSes.  Therefore, you can have multiple packets with the same timestamp even under a "light" traffic load.  For our purposes, I don't think it's a big deal, as the framework side has a TODO comment in meterpreter's sniffer.rb file that we should, at some point, add code to reorder packets based on packet ID in the event the sniffer has sent framework over a captured packet buffer which had wrapped.  But that code was never implemented, and even if we do implement it, the PktGetId as implemented here should be fine because framework will be looking for a **large** jump backwards between the IDs of two consecutive captured packets to determine "Ah, this is a buffer of capture packets which wrapped during capture, I need to reorder so the lowest ID is first in the pcap I'm generating..." and does the right thing.

That said, the request_sniffer_capture_dump() function in this file makes no attempt to start it's dump-to-framework of captured packets with the lowest-ID'd (e.g. oldest) packet in the buffer, rather it just starts at the lowest address of the packet capture buffer and goes from there.  This is *fine* if the capture buffer has never wrapped, but it totally looks like that if the capture buffer *has wrapped*, then when you go to dump the captured packets that this code will send over a block of packets which were "wrapped" to the beginning of the capture buffer, followed by a block of the oldest packets in the buffer.  And because the receiving framework code is lacking the reordering code ("TODO"), the generated pcap will have the packets in the wrong order, I believe.  If this is indeed the case, we could fix it either by adding the reordering code on the framework side OR by changing the dump code here to start dumping at the oldest packet in the buffer.